app/src/components/data-grid/autocomplete/AsyncAutocompleteDataGridEditCell.tsx

```diff
@@ -0,0 +1,204 @@
+import { DebouncedFunc } from 'lodash-es';
+import { useEffect, useState } from 'react';
+
+export interface IAsyncAutocompleteDataGridEditCell<
+  DataGridType extends GridValidRowModel,
+  ValueType extends string | number
+> {
+  /**
+   * Data grid props for the cell.
+   *
+   * @type {GridRenderCellParams<DataGridType>}
+   * @memberof IAsyncAutocompleteDataGridEditCell
+   */
+  dataGridProps: GridRenderCellParams<DataGridType>;
+  /**
+   * Function that returns a single option. Used to translate an existing value to its matching option.
+   *
+   * @memberof IAsyncAutocompleteDataGridEditCell
+   */
+  getCurrentOption: (value: ValueType) => Promise<IAutocompleteDataGridOption<ValueType> | null>;
+  /**
+   * Search function that returns an array of options to choose from.
+   *
+   * @memberof IAsyncAutocompleteDataGridEditCell
+   */
+  getOptions: DebouncedFunc<
+    (
+      searchTerm: string,
+      onSearchResults: (searchResults: IAutocompleteDataGridOption<ValueType>[]) => void
+    ) => Promise<void>
+  >;
+}
+
+/**
+ * Data grid single value asynchronous autocomplete component for edit.
+ *
+ * @template DataGridType
+ * @template ValueType
+ * @param {IAsyncAutocompleteDataGridEditCell<DataGridType, ValueType>} props
+ * @return {*}
+ */
+const AsyncAutocompleteDataGridEditCell = <DataGridType extends GridValidRowModel, ValueType extends string | number>(
+  props: IAsyncAutocompleteDataGridEditCell<DataGridType, ValueType>
+) => {
+  const { dataGridProps, getCurrentOption, getOptions } = props;
+
+  const apiRef = useGridApiContext();
+
+  // The current data grid value
+  const dataGridValue = dataGridProps.value;
+  // The input field value
+  const [inputValue, setInputValue] = useState<IAutocompleteDataGridOption<ValueType>['label']>('');
+  // The currently selected option
+  const [currentOption, setCurrentOption] = useState<IAutocompleteDataGridOption<ValueType> | null>(null);
+  // The array of options to choose from
+  const [options, setOptions] = useState<IAutocompleteDataGridOption<ValueType>[]>([]);
+  // Is control loading (search in progress)
+  const [isLoading, setIsLoading] = useState(false);
+
+  useEffect(() => {
+    let mounted = true;
+
+    if (!dataGridValue) {
+      // No current value
+      return;
+    }
+
+    if (dataGridValue === currentOption?.value) {
+      // Existing value matches tracked value
+      return;
+    }
+
+    const fetchCurrentOption = async () => {
+      // Fetch a single option for the current value
+      const response = await getCurrentOption(dataGridValue);
+
+      if (!mounted) {
+        return;
+      }
+
+      if (!response) {
+        return;
+      }
+
+      setCurrentOption(response);
+    };
+
+    fetchCurrentOption();
+
+    return () => {
+      mounted = false;
+    };
+  }, [dataGridValue, currentOption?.value, getCurrentOption]);
+
+  useEffect(() => {
+    let mounted = true;
```

Comment:
Possibly, though it would really just abstract away the initialization of the mounted variable, and the return at the end of the useEffect. You'd still have to check mounted before calling state hooks yourself.